### PR TITLE
Add WSGI web server, socket listen, fix bugs

### PR DESCRIPTION
**Pull request by xorbit**
- Added WSGI web server with up to 7 simultaneous connections, which
  are serviced sequentially
- adafruit_wiznet5k_wsgiserver based on and has interface identical to
  adafruit_esp32spi_wsgiserver for maximum compatibility
- Added socket 'listen' function
- Fixed socket 'connected' logic and made compatible with 'listen'
- Fixed source port > 65535 bug
- Removed SOCKETS list, it was buggy and caused get_socket allocation
  to only work if socket connections ended in order, which is not the
  case for web servers (and wrong in many other cases)
- Fixed TX buffer pointer limit to 16-bit
- Allow 5s for DHCP response

### CODE REVIEW

**Comment by brentru** [comment]
Woah...awesome! I'm going to review this tomorrow.

**Comment by tannewt** [comment]
Is it possible to split out the wsgi server bit to work with any socket implementation? That way we'd be able to share it across all network providers.

**Review by brentru — CHANGES_REQUESTED** [review]
Thanks for submitting this pull request. I have some questions along with some feedback.

**Comment by xorbit** [comment]
> Is it possible to split out the wsgi server bit to work with any socket implementation? That way we'd be able to share it across all network providers.

I'm afraid that may not be possible, at least as far as I understand things.  If you look at the [ESP32 implementation for instance](https://github.com/adafruit/Adafruit_CircuitPython_ESP32SPI/blob/master/adafruit_esp32spi/adafruit_esp32spi_wsgiserver.py), it's very different from what I had to do here for the W5500.  The ESP32 code has a "start server" call on the interface to listen on port 80, no socket involved at that point.  I had to do a weird non-standard thing where I listen on port 80 on several sockets at the same time, something that would not be possible on a standard socket the way I understand things.  But it was the only way I found to hold several connections at once until they can get served, instead of refusing connections after the first one is being served.

I think a common interface at the WSGI level is as good as it's going to get.  These chips all have different ways to do things and idiosyncrasies that may not be able to be resolved at the socket level.  If you want to investigates ways to make that happen, be my guest, but it seems like it would be a huge undertaking, likely invalidated the next time another chip is added.

File: adafruit_wiznet5k/adafruit_wiznet5k.py
Comment (xorbit) [reply adafruit_wiznet5k/adafruit_wiznet5k.py:590]: > Changed in version 3.5: The backlog parameter is now optional.

Best case this should be done in `adafruit_wiznet5k_socket.py`, not in the low level interface you specify.  How is the port number we listen to specified?  This (trying to be standards compliant in software on a limited platform) is not my usual playground, I'm more of a hardware guy.

File: adafruit_wiznet5k/adafruit_wiznet5k_socket.py
Comment (xorbit) [reply adafruit_wiznet5k/adafruit_wiznet5k_socket.py:68]: Why?  And what's supposed to be the interface for this?
The existing `SOCKET` list implementation didn't really do this and didn't in any way prevent sockets that can't exist from being allocated.

The constructor in `adafruit_wiznet5k_socket.py` uses the low level interface's `get_socket`, this just returns 0 if there is no free socket.  But 0 is a valid socket number and all kinds of ugly things happen when you use that afterward (like it may be stomping on the communication of the real owner of socket 0).

This is the way it was, and I don't see a reason to burden this pull request with having to fix everything in this driver.  To be honest, it was (and still is) an ugly and buggy piece of code, and it's going to take more than just this effort to fix it.  I just fixed some bugs I ran into in the process of implementing the server functionality, this is not a general big fix effort.

File: adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py
Comment (xorbit) [reply adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py:104]: Not really necessary I guess, it's something that happened in the development process.  Is there a spec that says how long a DHCP client should wait for a response?

File: adafruit_wiznet5k/adafruit_wiznet5k_socket.py
Comment (xorbit) [reply adafruit_wiznet5k/adafruit_wiznet5k_socket.py:165]: Except, the chip doesn't offer those calls of course.  Bind and listen could be faked (set the port with `bind`, then use it in the `listen` call), but the chip has no concept of `accept` where you get a new socket to communicate on.  It just considers the socket connected at that point and talks over it.  That's why I have to do the weird thing listening on the same port on several sockets at once.

File: adafruit_wiznet5k/adafruit_wiznet5k_wsgiserver.py
Comment (xorbit) [reply adafruit_wiznet5k/adafruit_wiznet5k_wsgiserver.py:1]: See my response below.

**Comment by tannewt** [comment]
👍 Thanks @xorbit. I wanted to make sure and chime in for awareness. Unifying things is my challenge to @brentru :-)

File: adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py
Comment (brentru) [reply adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py:104]: Ok, I looked deeper. Seems that [RFC3315](https://tools.ietf.org/html/rfc3315) specifies a timeout of 1s (**18.1.1. Creation and Transmission of Request Messages**
). Users should be able to pass `dhcp_timeout` to a `WIZNET5K` object's init if they need a longer response timeout or pass a timeout value to `set_dhcp`

File: adafruit_wiznet5k/adafruit_wiznet5k_socket.py
Comment (brentru) [reply adafruit_wiznet5k/adafruit_wiznet5k_socket.py]: I'm ok without `accept`.  It seems you implemented `bind` and `listen` in https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/pull/28/commits/a465bb0f01b890fb65c868108aa01627d92e2269, would you like me to re-review?

File: adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py
Comment (xorbit) [reply adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py:104]: That seems to be for the initial.  The maximum seems to be 30s according to 5.5?
I'm pretty sure 1s would fail on my network.

Comment (brentru) [reply adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py:104]: That seems correct - Could you increase the timeout to 30s here and within `adafruit_wiznet5k` (https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/blob/master/adafruit_wiznet5k/adafruit_wiznet5k.py#L164)?

**Comment by xorbit** [comment]
Not sure what the automatic check is complaining about, I don't think I touched the files that fail?

File: adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py
Comment (xorbit) [reply adafruit_wiznet5k/adafruit_wiznet5k_dhcp.py]: Changed the DHCP timeout to 30s.

**Comment by brentru** [comment]
@xorbit  looks like it's failing on 1) conflicting files and 2) a licensing issue.

@dherrada  - could you please take a look at the CI on this? Thanks!

File: adafruit_wiznet5k/adafruit_wiznet5k_socket.py
Comment (brentru) [reply adafruit_wiznet5k/adafruit_wiznet5k_socket.py]: It does - thanks for adding them. I'm going to test this on hardware before Friday :)

**Comment by dherrada** [comment]
@brentru On it. Probably just gonna copy the license files over from master since that seems like an easier way to do it.

**Comment by dherrada** [comment]
Uhh, so I'm not sure why but now the CI isn't running at all. The only merge conflict left is one line in the license of adafruit_wiznet5k.py. 
Screenshot from 2021-01-14 11-01-53

@brentru I'm not really experienced enough with git to know exactly what the right thing to do here is

**Comment by xorbit** [comment]
I did wonder about the license.  The license file and README both call out LGPL but the sources files themselves say MIT.  I did not make any change to this, but did wonder about the conflict there?

**Comment by tannewt** [comment]
@dherrada The CI won't run if you have merge conflicts. You'll need to do a `git merge` locally.

**Comment by dherrada** [comment]
@tannewt Ah, that makes sense. Good to know

**Comment by AdamCummick** [comment]
Any updates on this? I had actually started working on the same functionality and have a couple things I would like to add after in this is approved.

**Review by brentru — APPROVED** [review]
Tested simpletest and the new simple web server example. Thank you for adding this feature and bug fixes, @xorbit !

**Comment by brentru** [comment]
@AdamCummick  I just merged and released. Please feel free to open a separate PR with new functionalities and features. Thanks!

**Comment by xorbit** [comment]
Thanks @brentru !

**Pull request merged**
